### PR TITLE
fix dist train reduce mode

### LINKED ISSUE #13041: 多机同步时如果reduce_strategy使用fluid.BuildStrategy.ReduceStrategy.Reduce报错
```
Traceback (most recent call last):
  File "dist_mnist.py", line 163, in <module>
    runtime_main(TestDistMnist)
  File "/work/weike/svn/baidu/paddle/test/cts_test/dist_base.py", line 243, in runtime_main
    model.run_trainer(endpoints, trainer_id, trainers, run_params)
  File "/work/weike/svn/baidu/paddle/test/cts_test/dist_base.py", line 176, in run_trainer
    build_strategy=build_strategy)
  File "/usr/local/lib/python2.7/dist-packages/paddle/fluid/parallel_executor.py", line 165, in __init__
    build_strategy, num_trainers, trainer_id)
paddle.fluid.core.EnforceNotMet: can not find the right place for rpc op: recv at [/paddle/paddle/fluid/framework/details/multi_devices_graph_pass.cc:822]
PaddlePaddle Call Stacks:
```

### PR DESCRIPTION
**Pull request by typhoonzero**
Fix https://github.com/PaddlePaddle/Paddle/issues/13041

In reduce-mode, RPC op's device place is determined by previous ops to form the op dependency: "reduce->send->[send_barrier]->recv->[recv_barrier]->broadcast", in implementation before, send op tries to calculate a device place by original grad var name like "xxx@GRAD", but in sync_mode+reduce mode, the grad var name will be "xxx@GRAD.trainer_n", this var's device place will be created by reduce->send path, when come to recv op, it will still find place for origin "xxx@GRAD" name which will result in -1.

This will be fixed by setting send/recv op's attribute `op_role_var` to the actual grad var name.

### CODE REVIEW

**Comment by jacquesqiao** [comment]
求记录一下错误原因

File: python/paddle/fluid/tests/unittests/test_dist_mnist.py [hunk 1]
```diff
@@ -37,10 +38,30 @@ def test_se_resnext(self):
...
+        self._use_reduce = False
 
     def test_se_resnext(self):
         self.check_with_place("dist_mnist.py", delta=200)
 
 
+# FIXME(typhoonzero): enable these tests once we have 4
+# 4 GPUs on CI machine, and the base class should be updated.
+#
+# class TestDistMnist2x2ReduceMode(TestDistBase):
```
Comment (jacquesqiao) [inline python/paddle/fluid/tests/unittests/test_dist_mnist.py:50]: why we need 4 GPU for this test?

Comment (typhoonzero) [reply python/paddle/fluid/tests/unittests/test_dist_mnist.py:50]: Reduce mode need at least 2GPUs per trainer.

Comment (panyx0718) [reply python/paddle/fluid/tests/unittests/test_dist_mnist.py:50]: @chengduoZH should reduce mode fallback correctly when there is only 1 gpu detected?

**Review by panyx0718 — APPROVED** [review]

**Review by jacquesqiao — APPROVED** [review]
LGTM!

**Pull request merged**